stl/inc/random

```diff
@@ -4571,20 +4575,15 @@ public:
         param_type(initializer_list<_Ty> _Ilist, _Fn _Func) : _Mypbase(_Noinit) {
             if (2 <= _Ilist.size()) {
                 _Bvec.assign(_Ilist);
+
+                for (size_t _Idx = 0; _Idx < _Bvec.size() - 1; ++_Idx) {
+                    this->_Pvec.push_back(_Func(_Ty{0.5} * (_Bvec[_Idx] + _Bvec[_Idx + 1])));
+                }
             } else { // default construct
                 _Bvec.push_back(0);
                 _Bvec.push_back(1);

```

Comment:
```suggestion
                _Bvec = {0, 1};
```

Reply:
I am not really sure about that. Doesnt this involve an allocation as we need a temporary?

Reply:
`initializer_list` is not a container. It is not expected to allocate for its temporaries.

Reply:
Or do you imagine temporary `vector` constructed just before assignment? No, `vector` has `operator=` that accepts `initializer_list` directly

Reply:
https://godbolt.org/z/M43E5M

Both are not really good, but `initializer_list` version is still better.
At least allocates just once (like with reserve) and never checks if needed to reallocate (so no unnecessary branches):
```asm
        lea     edx, QWORD PTR [rax+2]    ; rax was zero plus two, so rdx is two
        call    int * std::allocator<int>::allocate(unsigned __int64) ; std::allocator<int>::allocate
```


clang version with `push_back` looks especially dumb. clang substituted all variables (and avoided branches), but still decided to allocate one element, then allocate two:
```asm
        mov     edi, 4
        call    operator new(unsigned long)
        mov     r15, rax
        mov     dword ptr [rax], 0
        add     rax, 4
        mov     qword ptr [r14], r15
        mov     qword ptr [r14 + 8], rax
        mov     qword ptr [r14 + 16], rax
        mov     edi, 8
        call    operator new(unsigned long)
```

Reply:
Honestly, this screams for `boost::small_vector<int, 2>`, that what I would have requested _at work_ :-)

Reply:
> Honestly, this screams for `boost::small_vector<int, 2>`, that what I would have requested _at work_ :-)

That would be ABI breaking.

Reply:
Sure, I'm not seriously proposing it here. Including 3rd party to public headers looks not an option, and implementing small object optimization right here adds up to the complexity.